src/libaktualizr/package_manager/packagemanagerinterface.cc

```diff
@@ -182,13 +180,97 @@ TargetStatus PackageManagerInterface::verifyTarget(const Uptane::Target& target)
+  struct statvfs stvfsbuf {};
+  const int stat_res = statvfs(config.images_path.c_str(), &stvfsbuf);
+  if (stat_res < 0) {
+    LOG_WARNING << "Unable to read filesystem statistics: error code " << stat_res;
+    return true;
+  }
+  const uint64_t available_bytes = (static_cast<uint64_t>(stvfsbuf.f_bsize) * stvfsbuf.f_bavail);
+  const uint64_t reserved_bytes = 1 << 20;
+
+  if (required_bytes + reserved_bytes < available_bytes) {
+    return true;
+  } else {
+    LOG_ERROR << "Insufficient disk space available to download target! Required: " << required_bytes
+              << ", available: " << available_bytes << ", reserved: " << reserved_bytes;
+    return false;
+  }
+}
+
+boost::optional<std::pair<uintmax_t, std::string>> PackageManagerInterface::checkTargetFile(
+    const Uptane::Target& target) const {
+  std::string filename = storage_->getTargetFilename(target.filename());
+  if (!filename.empty()) {
+    auto path = config.images_path / filename;
+    if (boost::filesystem::exists(path)) {
+      return {{boost::filesystem::file_size(path), path.string()}};
+    }
+  }
+  return boost::none;
+}
+
+std::ifstream PackageManagerInterface::openTargetFile(const Uptane::Target& target) const {
+  auto file = checkTargetFile(target);
+  if (!file) {
+    throw std::runtime_error("File doesn't exist for target " + target.filename());
+  }
+  std::ifstream stream(file->second, std::ios::binary);
+  if (!stream.good()) {
+    throw std::runtime_error("Can't open file " + file->second);
+  }
+  return stream;
+}
+
+std::ofstream PackageManagerInterface::createTargetFile(const Uptane::Target& target) {
+  std::string filename = target.hashes()[0].HashString();
+  std::string filepath = (config.images_path / filename).string();
+  boost::filesystem::create_directories(config.images_path);
+  std::ofstream stream(filepath, std::ios::binary | std::ios::ate);
+  if (!stream.good()) {
+    throw std::runtime_error("Can't write to file " + filepath);
+  }
+  storage_->storeTargetFilename(target.filename(), filename);
+  return stream;
+}
+
+std::ofstream PackageManagerInterface::appendTargetFile(const Uptane::Target& target) {
+  auto file = checkTargetFile(target);
+  if (!file) {
+    throw std::runtime_error("File doesn't exist for target " + target.filename());
+  }
+  std::ofstream stream(file->second, std::ios::binary | std::ios::app);
+  if (!stream.good()) {
+    throw std::runtime_error("Can't open file " + file->second);
+  }
+  return stream;
+}
+
+void PackageManagerInterface::removeTargetFile(const Uptane::Target& target) {
+  auto file = checkTargetFile(target);
+  if (!file) {
+    throw std::runtime_error("File doesn't exist for target " + target.filename());
+  }
+  boost::filesystem::remove(file->second);
+  storage_->deleteTargetInfo(target.filename());
+}
+
+std::vector<Uptane::Target> PackageManagerInterface::getTargetFiles() {
+  std::vector<Uptane::Target> v;
+  auto names = storage_->getAllTargetNames();
```

Comment:
Why return just the names as strings from `getAllTargetNames` if we just construct fake Targets here?

Reply:
Do you mean why not construct fake targets in the storage method, or why construct them at all?

Reply:
Both, I guess. It looks like previously we were constructing Targets in SqlStorage::getTargetFiles() based on the hashes that were stored in the database. What's the benefit of losing that information?

Reply:
There is no benefit in storing this information. Those Targets were still fake and used only for bookkeeping, as the comment in API header states:

>   Get list of targets currently in storage. This is intended to be used with
>   DeleteStoredTarget and targets are not guaranteed to be verified and
 >  up-to-date with current metadata.

For bookkeeping, it is sufficient to have only a filename.
I also think it is cleaner to keep in Storage only that logic that is directly related to reading/storing necessary information from db.

Reply:
Fair enough, thanks for clarifying!